src/Components/Components/src/Routing/Router.cs

```diff
@@ -173,16 +164,35 @@ private void Refresh(bool isNavigationIntercepted)
             }
         }
 
+        private async Task RunOnNavigateAsync(string path, bool isNavigationIntercepted)
+        {
+            if (OnNavigateAsync == null)
+            {
+                Refresh(isNavigationIntercepted);
+                return;
+
+            }
+            if (Loading != null)
+            {
+                _renderHandle.Render(Loading);
```

Comment:
I think this will cause double-rendering in the case where `OnNavigateAsync` returns a pre-completed task (which would be most of the time if most pages don't require loading any new assemblies). As in, first it will render "Loading", then instantly render the new page.

There's an existing pattern we've been using for async lifecycle methods:

 * Call the lifecycle method without `await` and get back the `Task`
 * If the task's already completed, move on synchronously without any `await`
 * Or, if the task wasn't already completed, only then render the intermediate state and do an `await` before rendering the final state

An example of that is [here](https://github.com/dotnet/aspnetcore/blob/master/src/Components/Components/src/ComponentBase.cs#L229), although it looks more complicated there because more stuff is going on that you might not need to handle here.

Also we need to decide how to handle the "task cancelled" state. For our `ComponentBase` lifecycle methods we handle it by [explicitly suppressing the cancellation exception](https://github.com/dotnet/aspnetcore/blob/master/src/Components/Components/src/ComponentBase.cs#L247) because it's not desirable to end up with a global unhandled exception (terminating the circuit for Blazor Server) when there are legit cases for cancelling. For example, if a user navigates while an earlier navigation is in progress, the developer might want to cancel whatever async task was in flight for the earlier navigation.

Reply:
Also, a minor point: I realised that calling this parameter `Loading` is a bit over-opinionated. Maybe `Navigating` would be a more accurate name, since the router is happy for you to do anything you want during asynchronous navigation - you *might* be loading some data, or you might be doing anything else.

Reply:
> I think this will cause double-rendering in the case where OnNavigateAsync returns a pre-completed task (which would be most of the time if most pages don't require loading any new assemblies). As in, first it will render "Loading", then instantly render the new page.

Correct. We only want to display the loading render fragment when awaiting the `OnNavigate`.

> An example of that is here, although it looks more complicated there because more stuff is going on that you might not need to handle here.

I did take a look at those and made the same assumption about complexity. Although as we start to consider all the different edge cases for this, that complexity might be warranted.

Reply:
I agree with Steve here, we should check if the task has run to completion and only call `_renderHandle.Render(Loading)` in that situation`

Reply:
> I agree with Steve here, we should check if the task has run to completion and only call _renderHandle.Render(Loading) in that situation`

I think you mean only call `_renderHandle.Render(Loading)` if the task has not run to completion. If so, that makes sense.